crates/solver/src/settlement/settlement_encoder.rs

```diff
@@ -121,19 +137,92 @@ impl SettlementEncoder {
             .token_index(order.order_creation.buy_token)
             .expect("missing buy token with price");
 
+        let order_trade = OrderTrade {
+            trade: Trade {
+                order,
+                sell_token_index,
+                executed_amount,
+                scaled_unsubsidized_fee,
+            },
+            buy_token_index,
+        };
+        let execution = order_trade
+            .trade
+            .executed_amounts(*sell_price, *buy_price)
+            .context("impossible trade execution")?;
+
+        self.order_trades.push(order_trade);
+        Ok(execution)
+    }
+
+    // Fails if any used sell-token doesn't have a price.
+    pub fn add_liquidity_order_trade(
+        &mut self,
+        order: Order,
+        executed_amount: U256,
+        scaled_unsubsidized_fee: U256,
+    ) -> Result<TradeExecution> {
+        // For the encoding strategy of liquidity orders, the sell prices are taken from
+        // the uniform clearing price vector. Therefore, either there needs to be an existing price
+        // for the sell token in the uniform clearing prices or we have to create a new price entry,
+        // if the sell token price is not yet available
+        let (sell_token_index, sell_price) = match self.token_index(order.order_creation.sell_token)
+        {
+            Some(index) => {
+                let sell_price = self
+                    .clearing_prices
+                    .get(&order.order_creation.sell_token)
+                    .context("settlement missing sell token")?;
+                (index, *sell_price)
+            }
+            None => {
+                let sell_token = order.order_creation.sell_token;
+                let sell_price = order.order_creation.buy_amount;
+                self.tokens.push(sell_token);
+                self.clearing_prices.insert(sell_token, sell_price);
+                (self.tokens.len() - 1, sell_price)
+            }
+        };
+
+        // Liquidity orders are settled at their limit price. We set:
+        // buy_price = sell_price * order.sellAmount / order.buyAmount, where sell_price is given from uniform clearing prices
+
+        // Rounding error checks:
```

Comment:
One other solution is to always add two new entries and have a duplicated:
```
tokens = [...tokens, sell_token, buy_token];
prices = [...prices, buy_amount, sell_amount];
```

This way you avoid rounding errors at the cost of potentially slightly more calldata (64 bytes of mostly 0s - so around 500 - 700 additional gas).

Reply:
Yes, one would avoid rounding errors and have less complexity in the driver. But since we have the rounding errors for nearly all solutions, I decided for the gas efficiency. Hope this is okay.

Reply:
Yes of course, the suggestion wasn't blocking the approval.